In [85]:
# read data
import pandas as pd
import os

In [113]:
# read data
data_dir = os.getcwd()
file_path = os.path.join(data_dir,"Humanyze_Internal_Dataset_email.xlsx")
# p_types = {'UserID': 'int64',
#            'Team': 'str',
#            'Location': 'str',
#            'Manager?': 'str'}
# e_types = {'Date':'str',
#            'Sender': 'int64',
#            'Receiver': 'int64',
#            'Receiver type': 'str',
#            'Count': 'int64'}

participants = pd.read_excel(file_path, sheet_name = "ParticipantInfo")
email = pd.read_excel(file_path, sheet_name = "Daily E-mail")

In [103]:
# summary of participant data; check unique values in string to see if NA values exist; turn nan in Manager? to 'N'
participants.dtypes
participants.head()
participants.describe(include='all')

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>UserID</th>
      <th>Team</th>
      <th>Location</th>
      <th>Manager?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>38.000000</td>
      <td>38</td>
      <td>38</td>
      <td>10</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>NaN</td>
      <td>4</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>top</th>
      <td>NaN</td>
      <td>Business</td>
      <td>Boston</td>
      <td>Y</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>NaN</td>
      <td>10</td>
      <td>24</td>
      <td>10</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>236648.447368</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>std</th>
      <td>70257.710228</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>min</th>
      <td>53534.000000</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>176763.000000</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>260585.500000</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>291084.250000</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>max</th>
      <td>323238.000000</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

In [104]:
# summary of email data
email.dtypes
email.head()
email.describe(include='all')

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Date</th>
      <th>Sender</th>
      <th>Receiver</th>
      <th>Receiver type</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>1500</td>
      <td>1500.000000</td>
      <td>1500.000000</td>
      <td>1500</td>
      <td>1500.000000</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>5</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>3</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>top</th>
      <td>2017-08-15 00:00:00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>CC</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>341</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>755</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>first</th>
      <td>2017-08-12 00:00:00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>last</th>
      <td>2017-08-16 00:00:00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>NaN</td>
      <td>239431.654000</td>
      <td>236143.796667</td>
      <td>NaN</td>
      <td>11.199333</td>
    </tr>
    <tr>
      <th>std</th>
      <td>NaN</td>
      <td>68074.408901</td>
      <td>67966.783802</td>
      <td>NaN</td>
      <td>8.635584</td>
    </tr>
    <tr>
      <th>min</th>
      <td>NaN</td>
      <td>53534.000000</td>
      <td>53534.000000</td>
      <td>NaN</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>NaN</td>
      <td>182838.000000</td>
      <td>174738.000000</td>
      <td>NaN</td>
      <td>5.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>NaN</td>
      <td>260833.000000</td>
      <td>260338.000000</td>
      <td>NaN</td>
      <td>9.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>NaN</td>
      <td>292634.000000</td>
      <td>292633.000000</td>
      <td>NaN</td>
      <td>14.000000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>NaN</td>
      <td>323238.000000</td>
      <td>323238.000000</td>
      <td>NaN</td>
      <td>44.000000</td>
    </tr>
  </tbody>
</table>
</div>

In [128]:
## check data quality
# check missing values
participants.isnull().sum()
email.isnull().sum()

# tranform nan in 'Manager?' to 'N'
participants['Manager?'].fillna('N',inplace=True)
participants.isnull().sum()

UserID      0
Team        0
Location    0
Manager?    0
dtype: int64

Found missing value in 'Manager?' column in participant data, transform to 'N'; No missing value in email data.

In [129]:
# check duplicates
participants[participants.duplicated(['UserID','Team','Location'])].shape
email[email.duplicated(['Date','Sender','Receiver','Receiver type'], keep=False)].sort_values(['Date','Sender','Receiver'])

# found 161 duplicates in email data; for duplicates, keep only the last in data
email = email.drop_duplicates(['Date','Sender','Receiver','Receiver type'], keep='last')
email.describe(include='all')

array(['TO', 'CC', 'BCC'], dtype=object)

In [328]:
'''
summary of in-team communications for two locations
'''
def data_summary(data,location):
    subs = data[data['Location'] == location]
    
    # total number of teams, people, average number of team members
    num_teams = len(subs['Team'].unique())
    num_people = len(subs['UserID'].unique())
    text = 'There are ' + str(num_teams) + ' teams in ' + location + ", and " + \
           str(num_people) + ' employees in total. The average number of team members is ' + \
           str(num_people/num_teams) + '.'
    print(text)
    
    # total number of managers, teams they manage, and the number of managers for one team
    num_managers = len(subs[subs['Manager?'] == 'Y']['UserID'].unique())
    text2 = 'There are ' + str(num_managers) + ' managers in ' + location
    if subs[subs['Manager?'] == 'Y'].duplicated(['UserID']).sum() == 0:
        text2 += '. Each manager only manages one team. '
    else:
        num_managing = subs[subs['Manager?'] == 'Y'].duplicated(['UserID'], keep=False).shape[0]
        text2 += '. On average, each manager manages ' + str(num_managing) + ' teams. '
    print(text2)
    
    for team in subs['Team'].unique():
        num_mng = subs[(subs['Manager?'] == 'Y') & (subs.Team == team)].shape[0]
        if num_mng != 1:
            print('The ' + team + ' team has ' + str(num_mng) + ' managers.')
        else:
            print('The ' + team + ' team has 1 manager.')
    print('\n')
data_summary(participants,'Boston')
data_summary(participants,'Palo Alto')

There are 3 teams in Boston, and 24 employees in total. The average number of team members is 8.0.
There are 6 managers in Boston. Each manager only manages one team. 
The Business team has 2 managers.
The Tech DC team has 2 managers.
The Tech HYPE team has 2 managers.


There are 2 teams in Palo Alto, and 14 employees in total. The average number of team members is 7.0.
There are 4 managers in Palo Alto. Each manager only manages one team. 
The Business team has 1 manager.
The Analytics team has 3 managers.




In [327]:
# Task 1: using the data provided, verify if the CEO's hypotheses are true
# First define cohesion and manager visibility; measure cohesion by daily average email sent to teammates per person 
# in one location; measure manager visibility by daily average email communication
# (including sent and received) with manager per person in one location (not including communications between managers)

# check if one team's members are only in one location; verified
participants[['Team','Location']].drop_duplicates().shape[0]

# merge with email data
sub1 = pd.merge(email, participants, left_on='Sender', right_on="UserID", how='left')
total = pd.merge(sub1, participants, left_on='Receiver',right_on='UserID',how='left')
inteam_total = total[total.Team_x == total.Team_y]
between_team = total[total.Team_x != total.Team_y]

# 332 emails in team, 1086 between teams

def avg_email(location, between = False, by_team = False):
    if(between):
        subs = between_team[between_team.Location_x == location]
        text1 = ', the total number of emails sent from employees to their non-teammates (between-team communication) in 5 days is '
        text2 = ', the daily average between-team emails sent per person is '
    else:
        subs = inteam_total[inteam_total.Location_x == location]
        text1 = ', the total number of emails sent from employees to their teammates (in-team communication) in 5 days is '
        text2 = ', the daily average in-team emails sent per person is '
        
    num_email = sum(subs.Count)
    num_employee = len(participants[participants.Location == location]['UserID'].unique())
    avg_email = round(num_email/num_employee/5,2)

    text = 'In ' + location + text1 + str(num_email) + '. As the total number of employees in ' + \
           location + ' is ' + str(num_employee) + text2  + str(avg_email) + '.'
    print(text)
    
    if(by_team):
        group_sum = subs.groupby('Team_x')['Count'].sum()
        group_count = participants[participants.Location == location].groupby('Team')['UserID'].nunique()
        by_team_count = round(group_sum.divide(group_count)/5,2)
        by_team_total = pd.DataFrame(dict(Num_email_5_days = group_sum, 
                                          Num_employee = group_count, 
                                          Daily_Average_email = by_team_count)).reset_index()
        print('The total emails (5 days), number of employees and daily average emails sent for each team is:')
        print(by_team_total)
        print('\n')

avg_email('Boston', by_team=True)
avg_email('Palo Alto',by_team=True)      
avg_email('Boston', between=True, by_team=True)
avg_email('Palo Alto', between=True,by_team=True)


In Boston, the total number of emails sent from employees to their teammates (in-team communication) in 5 days is 4868. As the total number of employees in Boston is 24, the daily average in-team emails sent per person is 40.57.
The total emails (5 days), number of employees and daily average emails sent for each team is:
       index  Daily_Average_email  Num_email_5_days  Num_employee
0   Business                28.40               852             6
1    Tech DC                49.58              2231             9
2  Tech HYPE                39.67              1785             9


In Palo Alto, the total number of emails sent from employees to their teammates (in-team communication) in 5 days is 3124. As the total number of employees in Palo Alto is 14, the daily average in-team emails sent per person is 44.63.
The total emails (5 days), number of employees and daily average emails sent for each team is:
       index  Daily_Average_email  Num_email_5_days  Num_employee
0  Analytics  

Cohesion: The daily average number of in-team email sent per person at Boston office is 40.57, which is fewer than that of Palo Alto office (44.63). However, the daily average number of between-team email sent per person at Boston office is 46.44, which is much more than that of Palo Alto office (34.01).

In [326]:
# manager visibility
def manager_email(location, by_team = False):
    subs = inteam_total[(inteam_total.Location_x == location) & 
                        (inteam_total['Manager?_x'] != inteam_total['Manager?_y'])]
    num_email = sum(subs.Count)
    num_employee = len(participants[participants.Location == location]['UserID'].unique())
    num_managers = len(participants[(participants.Location == location) 
                                    & (participants['Manager?'] == 'Y')]['UserID'].unique())
    num_non_manager = num_employee - num_managers
    avg_email = round(num_email/num_non_manager/5,2)
    text = 'In ' + location + ', the total number of emails (sent and received) between employees and their managers in 5 days is ' + \
           str(num_email) + '. As the total number of non-manager employees in ' + \
           location + ' is ' + str(num_non_manager) + ', the daily average emails between employees and their managers is ' \
           + str(avg_email) + '.'
    print(text)
    
    if by_team:
        group_sum = subs.groupby('Team_x')['Count'].sum()
        sum_emp = participants[(participants['Manager?'] == 'N') & (participants.Location == location)].groupby('Team')['UserID'].nunique()
        avg = round(group_sum.divide(sum_emp)/5,2)
        by_team_total = pd.DataFrame(dict(Num_email_5_days = group_sum,Num_non_mng_employee = sum_emp,Daily_average_email = avg)).reset_index()
        print('The total emails (5 days), number of employees and the daily average emails between employees and their managers for each team is:')
        print(by_team_total)
        print('\n')
    
manager_email('Boston',by_team=True)
manager_email('Palo Alto',by_team=True)
                         

In Boston, the total number of emails (sent and received) between employees and their managers in 5 days is 2220. As the total number of non-manager employees in Boston is 18, the daily average emails between employees and their managers is 24.67.
The total emails (5 days), number of employees and the daily average emails between employees and their managers for each team is:
       index  Daily_average_email  Num_email_5_days  Num_non_mng_employee
0   Business                21.40               428                     4
1    Tech DC                31.11              1089                     7
2  Tech HYPE                20.09               703                     7


In Palo Alto, the total number of emails (sent and received) between employees and their managers in 5 days is 1486. As the total number of non-manager employees in Palo Alto is 10, the daily average emails between employees and their managers is 29.72.
The total emails (5 days), number of employees and the daily average 

Manager visibility: The daily average number of emails (sent and received) between employee and manager for each employee at Boston office is 24.67, which is fewer than that of Palo Alto office(29.72).

Thus, the CEO's hypotheses about the relationship between engagement survey score and in-team communications at the two offices, cannot be confirmed based on the data provided for now.

From above analysis we can see that considering different metrics (for example, in-team and between-team communications), the result of comparison between Boston office and Palo Alto office can be very different. Thus, we can next consider more metrics across longer time period, to see if any of them matches with the survey result. Also, further segmentation (for example, by team as above) can be conducted to see if any specific team account for the difference, or it is a general difference across all teams at the same location. Then if we can target specific teams, for those having significantly high or low level of engagement and score, we can try to find the reason and see if other teams can learn from them.

1. As the dataset contains only 5 days' data, the results might be by chance. For example, maybe for this week's one activity, there are more email communications in some specific teams at Palo Alto or Boston office.
2. Email is just one type of communication method, and the total/average number of emails is just one ofthe many measurements of engagement. The teams having less email communications may have more phone calls, live chats or team meetings.
3. Apart from in-team communication, between-team communication can also be an important factor influencing employee's engagement level, as mentioned in the above analysis.

More detailed about email communication:
1. Replying time/replying percentage can also be good measurements for interaction level.
2. The definition for cohesion and manager visibility might be different.